pkg/server/plugin/upstreamca/disk/disk.go

```diff
@@ -63,67 +67,86 @@ func (m *DiskPlugin) Configure(ctx context.Context, req *spi.ConfigureRequest) (
 		return nil, errors.New("trust_domain is required")
 	}
 
+	ttl, err := time.ParseDuration(config.TTL)
+	if err != nil {
+		return nil, fmt.Errorf("invalid TTL value: %v", err)
+	}
+
+	config.ttl = ttl
+	config.trustDomain = req.GlobalConfig.TrustDomain
+
+	upstreamCA, cert, err := p.loadUpstreamCAAndCert(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load upstream CA: %v", err)
+	}
+
+	// Set local vars from config struct
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	p.config = config
+	p.cert = cert
+	p.upstreamCA = upstreamCA
+
+	return &spi.ConfigureResponse{}, nil
+}
+
+func (p *DiskPlugin) loadUpstreamCAAndCert(config *Configuration) (*x509svid.UpstreamCA, *x509.Certificate, error) {
 	key, err := pemutil.LoadPrivateKey(config.KeyFilePath)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	cert, err := pemutil.LoadCertificate(config.CertFilePath)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	// Validate cert matches private key
 	matched, err := x509util.CertificateMatchesPrivateKey(cert, key)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 	if !matched {
-		return nil, errors.New("certificate and private key does not match")
+		return nil, nil, errors.New("certificate and private key does not match")
 	}
 
-	ttl, err := time.ParseDuration(config.TTL)
-	if err != nil {
-		return nil, fmt.Errorf("invalid TTL value: %v", err)
-	}
-
-	// Set local vars from config struct
-	m.mtx.Lock()
-	defer m.mtx.Unlock()
-
-	m.cert = cert
-	m.upstreamCA = x509svid.NewUpstreamCA(
+	return x509svid.NewUpstreamCA(
 		x509util.NewMemoryKeypair(cert, key),
-		req.GlobalConfig.TrustDomain,
+		config.trustDomain,
 		x509svid.UpstreamCAOptions{
-			SerialNumber: m.serialNumber,
-			TTL:          ttl,
-		})
-
-	return &spi.ConfigureResponse{}, nil
+			SerialNumber: p.serialNumber,
+			TTL:          config.ttl,
+		},
+	), cert, nil
 }
 
 func (*DiskPlugin) GetPluginInfo(context.Context, *spi.GetPluginInfoRequest) (*spi.GetPluginInfoResponse, error) {
 	return &spi.GetPluginInfoResponse{}, nil
 }
 
-func (m *DiskPlugin) SubmitCSR(ctx context.Context, request *upstreamca.SubmitCSRRequest) (*upstreamca.SubmitCSRResponse, error) {
-	m.mtx.RLock()
-	defer m.mtx.RUnlock()
+func (p *DiskPlugin) SubmitCSR(ctx context.Context, request *upstreamca.SubmitCSRRequest) (*upstreamca.SubmitCSRResponse, error) {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
 
-	if m.upstreamCA == nil {
-		return nil, errors.New("invalid state: not configured")
+	upstreamCA, upstreamCert, err := p.loadUpstreamCAAndCert(p.config)
+	if err != nil && p.upstreamCA == nil {
+		return nil, fmt.Errorf("no cached CA and failed to load CA: %v", err)
+	}
+	if err == nil {
+		p.upstreamCA = upstreamCA
```

Comment:
I see I misread an `Rlock` as a `Lock` :) Will fix.

Reply:
these need to be set under a write lock or it will race with concurrent callers of SubmitCSR. Maybe something like:

```
func (p *DiskPlugin) SubmitCSR(ctx context.Context, request *upstreamca.SubmitCSRRequest) (*upstreamca.SubmitCSRResponse, error) {
	upstreamCA, upstreamCert, err := p.reloadCA()
	if err != nil {
		return nil, err
	}

	cert, err := upstreamCA.SignCSR(...)

	// ... rest of the goodness ...
}

func (p *DiskPlugin) reloadCA() (*x509svid.UpstreamCA, *x509.Certificate, error) {
	p.mtx.Lock()
	defer p.mtx.Unlock()
	upstreamCA, upstreamCert, err := p.loadUpstreamCAAndCert(p.config)
	switch {
	case err == nil:
		p.upstreamCA = upstreamCA
		p.cert = upstreamCert
	case p.upstreamCA != nil:
		upstreamCA = p.upstreamCA
		upstreamCert = p.cert
	default:
		return nil, fmt.Errorf("no cached CA and failed to load CA: %v", err)
	}
	return upstreamCA, upstreamCert, nil
}
```